drivers/SmartThings/zigbee-switch/src/zll-dimmer-bulb/ikea-xy-color-bulb/init.lua

```diff
@@ -64,38 +67,87 @@ local query_device = function(device)
 end
 
 local function set_color_handler(driver, device, cmd)
-  local hue = cmd.args.color.hue > 99 and 99 or cmd.args.color.hue
+  -- Cancel the hue/sat timer if it's running, since setColor includes both hue and saturation
+  local huesat_timer = device:get_field(HUESAT_TIMER)
+  if huesat_timer ~= nil then
+    device.thread:cancel_timer(huesat_timer)
+    device:set_field(HUESAT_TIMER, nil)
+  end
+
+  local hue = (cmd.args.color.hue ~= nil and cmd.args.color.hue > 99) and 99 or cmd.args.color.hue
   local sat = cmd.args.color.saturation
+
   local x, y, Y = utils.safe_hsv_to_xy(hue, sat)
   store_xyY_values(device, x, y, Y)
-  switch_defaults.on(driver,device,cmd)
+  switch_defaults.on(driver, device, cmd)
 
   device:send(ColorControl.commands.MoveToColor(device, x, y, 0x0000))
 
+  device:set_field(TARGET_HUE, nil)
+  device:set_field(TARGET_SAT, nil)
   device.thread:call_with_delay(2, query_device(device))
 end
 
-local function set_hue_handler(driver, device, cmd)
-  local sat = device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.saturation.NAME)
-  local hue = cmd.args.hue > 99 and 99 or cmd.args.hue
-  local x, y, Y = utils.safe_hsv_to_xy(hue, sat)
-  store_xyY_values(device, x, y, Y)
-  switch_defaults.on(driver,device,cmd)
-
-  device:send(ColorControl.commands.MoveToColor(device, x, y, 0x0000))
+local huesat_timer_callback = function(driver, device, cmd)
+  return function()
+    local hue = device:get_field(TARGET_HUE)
+    local sat = device:get_field(TARGET_SAT)
+    hue = hue ~= nil and hue or device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.hue.NAME)
+    sat = sat ~= nil and sat or device:get_latest_state("main", capabilities.colorControl.ID, capabilities.colorControl.saturation.NAME)
+    cmd.args = {
+      color = {
+        hue = hue,
+        saturation = sat
+      }
+    }
+    set_color_handler(driver, device, cmd)
+  end
+end
 
-  device.thread:call_with_delay(2, query_device(device))
+local function set_hue_handler(driver, device, cmd)
+  local hue = cmd.args.hue
+  local huesat_timer = device:get_field(HUESAT_TIMER)
+  if huesat_timer ~= nil then
+    device.thread:cancel_timer(huesat_timer)
+    device:set_field(HUESAT_TIMER, nil)
+    local sat = device:get_field(TARGET_SAT)
+    if sat ~= nil then
+      cmd.args = {
+        color = {
+          hue = hue,
+          saturation = sat
+        }
+      }
+      set_color_handler(driver, device, cmd)
+      return
```

Comment:
multiple exit points tsk tsk

Reply:
That's a fair nit, plus there is duplicate code. Take a look at the fixup commit and see if you think this is better.